Dataset on student exam score


In [ ]:
import pandas as pd

data = {
    "Study_Hours":[2,3,4,5,6,7,8,9,10,11],
    "Actual":[50,55,65,70,75,80,85,90,95,98],
    "Predicted":[48,54,63,72,74,79,83,88,92,96]
}

df = pd.DataFrame(data)

df

,Study_Hours,Actual,Predicted
0,2,50,48
1,3,55,54
2,4,65,63
3,5,70,72
4,6,75,74
5,7,80,79
6,8,85,83
7,9,90,88
8,10,95,92
9,11,98,96


In [ ]:
from sklearn.metrics import mean_absolute_error , mean_squared_error , root_mean_squared_error , mean_absolute_percentage_error , r2_score

mae = mean_absolute_error(df["Actual"], df["Predicted"])
print("MAE:", mae)

# formula MAE
# MAE=1/n ​∑ ∣y−y'​∣
# absolute error = 2,1,2,2,1,1,2,2,3,2
# sum = 18
# MAE = 18/10
# MAE = 1.8

MAE: 1.8


In [ ]:
mse = mean_squared_error(df["Actual"], df["Predicted"])
print("MSE:", mse)

# formula MSE
# MSE=1/n ​∑ (y−y'​)^2
# squaring absolute error = 4,1,4,4,1,1,4,4,9,4
# sum = 36
# MSE = 36/10
# MSE = 3.6

MSE: 3.6


In [ ]:
rmse = root_mean_squared_error(df['Actual'], df["Predicted"])
print("RMSE:",rmse)

# formula RMSE
# RMSE= \sqrt{1/n ​∑ (yi-yi')^2}
# squaring absolute error = 4,1,4,4,1,1,4,4,9,4
# sum = 36
# 1/n ​∑ = 36/10
# /sqrt 3.6
# RMSE = 1.89


RMSE: 1.8973665961010275


In [ ]:
mape = mean_absolute_percentage_error(df["Actual"], df["Predicted"])
print("MAPE:",mape)

# formula MAPE
# MAPE= 100/n ​∑ |(yi-yi')/yi|
# |(yi-yi')/yi| = 0.040 , 0.018 , 0.030 , 0.028 , 0.013 , 0.012 , 0.023 , 0.022 , 0.031 , 0.020
# ​∑ |(yi-yi')/yi| = 0.237
# MAPE = 100/10 * 0.237
# MAPE = 2.37 %

MAPE: 0.02410945554764661


In [ ]:
r2 = r2_score(df["Actual"], df["Predicted"])
print("R2:", r2)

# formula R2
# R2=1−SSR/SST
# SSR=∑(y−y'​)^2​
# SST=∑(y−yˉ​)^2
# yˉ=763/10
# yˉ=76.3
# SSR=36
# SST = (y-yˉ)^2 = 691.69 , 453.69 , 127.69 , 39.69 , 1.69 , 13.69 , 75.69 , 187.69 , 349.69 , 470.89
# SST = ∑(y−yˉ​)^2
# SST = 2412.10
# R2 = 1-SSR/SST
# R2 = 1-36/2412.10
# R2 = 0.985

R2: 0.9850752456365822


In [ ]:
n = len(df)
k = 1

adj_r2 = 1 - (1 - r2)*(n - 1)/(n - k - 1)

print("Adjusted R2:", adj_r2)


Adjusted R2: 0.983209651341155


In [ ]:
print("ALL REGRESSION METRICES VALUES: ")
print("MAE:", mae)
print("MSE:", mse)
print("RMSE:", rmse)
print("MAPE:", mape)
print("R2:", r2)
print("Adjusted R2:", adj_r2)

ALL REGRESSION METRICES VALUES: 
MAE: 1.8
MSE: 3.6
RMSE: 1.8973665961010275
MAPE: 0.02410945554764661
R2: 0.9850752456365822
Adjusted R2: 0.983209651341155


INTERPRETATION FROM THE METRICES

**MAE : 1.8**

MAE tells the average absolute error between actual and predicted values. On average, the predicted values differ from the actual values by about 1.8 units.

Since the scores range roughly 50 - 98, an error of 1.8 is very small, indicating good prediction accuracy.


**MSE = 3.6**

MSE measures average squared prediction errors.
Because errors are squared, large mistakes are penalized more.

A value of 3.6 is low, meaning the model does not make large prediction errors.

**RMSE = 1.90**

RMSE represents the typical size of prediction errors.
Predictions are typically about 1.9 units away from the actual values.

**MAPE = 2.37%**

MAPE measures average prediction error in percentage.
Predictions differ from actual values by about 2.37% on average.

**R² = 0.985**

How much of the variation in the target variable(ACTUAL SCORE) is explained by the model.The values change from 50 to 98, so there is variation in the data.
Model uses Study Hours to predict the score.

98.5% of the variation in scores is explained by study hours.

**Adjusted R² = 0.983**

Adjusted R² corrects R² for number of predictors.

Adding more variables always increases R²

Even if those variables are useless

Adjusted R² asks :
Did the predictor actually improve the model?

After accounting for the number of predictors, the model still explains 98.3% of the variation in the data.



